bigquery/storage_iterator.go

```diff
@@ -0,0 +1,267 @@
+	return it, nil
+}
+
+func newStorageRowIteratorFromJob(ctx context.Context, job *Job, totalRows uint64) (*RowIterator, error) {
+	cfg, err := job.Config()
+	if err != nil {
+		return nil, err
+	}
+	qcfg := cfg.(*QueryConfig)
+	if qcfg.Dst == nil {
+		// TODO: script job ?
+		return nil, fmt.Errorf("nil job destination table")
+	}
+	return newStorageRowIteratorFromTable(ctx, qcfg.Dst)
+}
+
+func newRawStorageRowIterator(ctx context.Context, rs *readSession) (*arrowIterator, error) {
+	arrowIt := &arrowIterator{
+		ctx:     ctx,
+		session: rs,
+		records: make(chan arrowRecordBatch, rs.rc.maxWorkerCount+1),
+		errs:    make(chan error, rs.rc.maxWorkerCount+1),
+	}
+	if rs.bqSession == nil {
+		err := rs.start()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return arrowIt, nil
+}
+
+func newStorageRowIterator(ctx context.Context, rs *readSession, totalRows uint64) (*RowIterator, error) {
+	arrowIt, err := newRawStorageRowIterator(ctx, rs)
+	if err != nil {
+		return nil, err
+	}
+	it := &RowIterator{
+		ctx:           ctx,
+		arrowIterator: arrowIt,
+		TotalRows:     totalRows,
+		rows:          [][]Value{},
+	}
+	it.nextFunc = nextFuncForStorageIterator(it)
+	it.pageInfo = &iterator.PageInfo{
+		Token:   "",
+		MaxSize: int(totalRows),
+	}
+	return it, nil
+}
+
+func nextFuncForStorageIterator(it *RowIterator) func() error {
+	return func() error {
+		if len(it.rows) > 0 {
+			return nil
+		}
+		arrowIt := it.arrowIterator
+		record, err := arrowIt.next()
+		if err == iterator.Done {
+			if len(it.rows) == 0 {
+				return iterator.Done
+			}
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+
+		rows, err := arrowIt.decoder.decodeArrowRecords(record)
+		if err != nil {
+			return err
+		}
+		it.rows = rows
+		return nil
+	}
+}
+
+func (it *arrowIterator) init() error {
+	if it.decoder != nil { // Already initialized
+		return nil
+	}
+
+	bqSession := it.session.bqSession
+	if bqSession == nil {
+		return errors.New("read session not initialized")
+	}
+
+	streams := bqSession.Streams
+	if len(streams) == 0 {
+		return iterator.Done
+	}
+
+	if it.schema == nil {
+		meta, err := it.session.table.Metadata(it.ctx)
+		if err != nil {
+			return err
+		}
+		it.schema = meta.Schema
+	}
+
+	decoder, err := newArrowDecoderFromSession(it.session, it.schema)
+	if err != nil {
+		return err
+	}
+	it.decoder = decoder
+
+	wg := sync.WaitGroup{}
+	sem := semaphore.NewWeighted(int64(it.session.rc.maxWorkerCount))
+	go func() {
+		wg.Wait()
+		close(it.records)
+		close(it.errs)
+		it.done = true
+	}()
+
+	go func() {
+		for _, readStream := range streams {
+			wg.Add(1)
+			err := sem.Acquire(it.ctx, 1)
+			if err != nil {
+				it.errs <- err
+				sem.Release(1)
+				wg.Done()
+				break
+			}
+			go func(readStreamName string) {
+				it.processStream(readStreamName)
+				sem.Release(1)
+				wg.Done()
+			}(readStream.Name)
+		}
+	}()
+	return nil
+}
+
+func (it *arrowIterator) processStream(readStream string) {
```

Comment:
Should this (or consumeRowStream) allow for retries before propagating error?

Reply:
Current implementation is appropriate for a unary call, but could be refined to something more stream oriented (e.g. fail after N consecutive attempts, reset attempts when we successfully request).